Converting BasicEngine API to TFLite API

manderfeld opened the pull request:
This PR converts all examples for Gstreamer, OpenCV, PyGame, and Raspicam. The README files and install_requirements.sh scripts were also updated as needed.

The make_interpreter, input_size, input_tensor, output_tensor, set_input, get_output functions are all based on [this TFLite classification example](https://github.com/google-coral/tflite/blob/master/python/examples/classification/classify_image.py) whose functions are defined [here](https://github.com/google-coral/tflite/blob/master/python/examples/classification/classify.py)

Tested: Camera examples run for the Coral Dev Board, Raspberry Pi with USB Accelerator, and Linux with USB Accelerator. 

In open_cv/install_requirements.sh for Raspberry Pi when if it just calls ```sudo pip3 install opencv-contrib-python``` an error states that it’s unable to locate package python3-opencv. This works for the Coral Board (and Linux). The only instructions I found for installing OpenCV for Raspberry Pi are a [very involved process](https://www.deciphertechnic.com/install-opencv-python-on-raspberry-pi/) (I could still try to include this in the script), but the other fix I found uses opencv-contrib-python and installs some additional libraries which I added to the install_requirements.sh script

manderfeld commented on opencv/install_requirements.sh:
```diff
@@ -23,4 +23,6 @@ if grep -s -q "MX8MQ" /sys/firmware/devicetree/base/model; then
 fi
 
-sudo apt install python3-opencv
+sudo pip3 install opencv-contrib-python
```
@mtyka FYI this is an unofficial OpenCV package (it's a pre-built version of OpenCV),  but I could also modify to script to do so, it's just a [very involved process](https://www.deciphertechnic.com/install-opencv-python-on-raspberry-pi/)

mtyka replied:
I think this package is only available starting from  Raspbian Buster, the latest version of Raspbian. I just tried it and it work. 
Thus think we should stick with "sudo apt install python3-opencv".
If the unofficial one works with Raspbian Stretch then perhaps just leave a comment here to  that effect.

amj commented on gstreamer/common.py:
```diff
@@ -14,9 +14,47 @@
+        input_tensor(interpreter)[:, :] = np_buffer
+        buf.unmap(mapinfo)
+
+def set_interpreter(interpreter, data):
```
needs a docstring.  To an unfamiliar reader, it looks like it sets an input, not an interpreter, so a comment would be helpful.

amj commented on gstreamer/classify.py:
```diff
@@ -51,27 +72,29 @@ def main():
-      results = engine.classify_with_input_tensor(input_tensor,
-          threshold=args.threshold, top_k=args.top_k)
+      common.set_interpreter(interpreter, input_tensor)
+      # For larger input tensor sizes, use the edgetpu.classification.engine for better performance
```
larger batches, larger image sizes, or both?

manderfeld replied:
Jonas suggested a comment saying that "for larger input tensors one API is currently a bit faster than the other" (that the old API would perform better than TFLite). To my knowledge this would be at least larger image sizes, what do you mean by batches?

Thanks!

amj replied:
https://github.com/google-coral/examples-camera/blob/d1ddb17d0a2c259089e1ce4d2768954492b1f67e/pygame/classify_capture.py#L43 .  I assumed the _ here was for the number of batches, i.e., classifying more than one image at a time.

Coral i think is better for smaller batch sizes?  But i'm not sure, which is why i was asking :)

manderfeld replied:
I'll change the comment to say 'For larger input image sizes..." for clarity

amj commented on gstreamer/classify.py:
```diff
@@ -35,6 +39,23 @@ def generate_svg(size, text_lines):
+        for i in np.argpartition(scores, -top_k)[-top_k:]
+        if scores[i] >= score_threshold
+    ]
+    return sorted(classes, key=operator.itemgetter(1), reverse=True)
```
usually i see key funcs as lambda's, e.g. `key lambda i: i[1]`.  Is using the `operator` module a local idiom?

manderfeld replied:
For get_output I referenced [this TFLite example ](https://github.com/google-coral/tflite/blob/master/python/examples/classification/classify.py#L49) and they use the `operator` module here. Thoughts?

amj replied:
a quick google suggests its slightly faster! https://stackoverflow.com/questions/17243620/operator-itemgetter-or-lambda

mtyka replied:
+1 to operator.itemgetter() though it really doesn't matter in practice here.

amj commented on gstreamer/detect.py:
```diff
@@ -54,7 +57,7 @@ def generate_svg(src_size, inference_size, inference_box, objs, labels, text_lin
         shadow_text(dwg, 10, y*20, line)
     for obj in objs:
-        x0, y0, x1, y1 = obj.bounding_box.flatten().tolist()
+        x0, y0, x1, y1 = list(obj.bbox)
```
:+1:

amj commented on gstreamer/detect.py:
```diff
@@ -64,12 +67,41 @@ def generate_svg(src_size, inference_size, inference_box, objs, labels, text_lin
         x, y, w, h = x * scale_x, y * scale_y, w * scale_x, h * scale_y
         percent = int(100 * obj.score)
-        label = '%d%% %s' % (percent, labels[obj.label_id])
+        label = '%d%% %s' % (percent, labels.get(obj.id, obj.id))
```
nit: consider using `"".format()` here for easier human-parsing :)

amj commented on gstreamer/detect.py:
```diff
@@ -86,23 +118,22 @@ def main():
-                                    threshold=args.threshold,
-                                    top_k=args.top_k)
+      common.set_interpreter(interpreter, input_tensor)
+      # For larger input tensor sizes, use the edgetpu.detection.engine for better performance
```
as above.

amj commented on pygame/README.md:
```diff
@@ -3,10 +3,17 @@
+    device](https://coral.ai/docs/setup/). If it's been a while, repeat to be sure
+    you have the latest software.
+
+    Importantly, you should have the latest TensorFlow Lite runtime installed
```
Might be worth adding "you can check which version is installed with `pip3 freeze`" or similar, i dunno.

manderfeld replied:
Would there be a way to do this that wouldn't involve hardcoding the most recent version of tflite-runtime or having the user also go to the [Python quickstart guide](https://www.tensorflow.org/lite/guide/python) (the page where I could clearly find version information)?

amj replied:
I don't think you need to find a way to pull the expected version and keep the doc updated, i'm saying just telling them (perhaps a beginner), how to quickly check the version they have ( `pip freeze` ) would make it beginner friendly.

mtyka replied:
```pip3 show tflite_runtime``` will show you the version. Currently we require 1.14 i think

amj commented on pygame/classify_capture.py:
```diff
@@ -14,16 +14,61 @@
+import tflite_runtime.interpreter as tflite
+import time
+
+Class = collections.namedtuple('Class', ['id', 'score'])
```
nit: how about "Category"?  Or "Classification" ?

mtyka replied:
Agree "Class" has overloaded meanings in programming languages, and "class" is a reserved keyword.
Unfortunately our official tflite examples also use "classes", "Class" etc. which is unfortunate. 
I agree lets go to "category" or "classification"

amj commented on pygame/classify_capture.py:
```diff
@@ -14,16 +14,61 @@
+        for i in np.argpartition(scores, -top_k)[-top_k:]
+        if scores[i] >= score_threshold
+    ]
+    return sorted(classes, key=operator.itemgetter(1), reverse=True)
```
as above.

mtyka replied:
(reg class -> category)

amj commented on opencv/install_requirements.sh:
```diff
@@ -23,4 +23,6 @@ if grep -s -q "MX8MQ" /sys/firmware/devicetree/base/model; then
 fi
 
-sudo apt install python3-opencv
+sudo pip3 install opencv-contrib-python
```
@mtyka any problem with adding these dependencies?

amj commented on pygame/detect.py:
```diff
@@ -85,17 +143,18 @@ def main():
                rect = pygame.Rect(x0 * cam_w, y0 * cam_h, (x1 - x0) * cam_w, (y1 - y0) * cam_h)
                pygame.draw.rect(mysurface, red, rect, 1)
-               label = "%.0f%% %s" % (100*result.score, labels[result.label_id])
+               label = "%.0f%% %s" % (100*result.score, labels.get(result.id, result.id))
```
nit: consider "".format() here

amj commented on raspicam/classify_capture.py:
```diff
@@ -14,14 +14,59 @@
+        for i in np.argpartition(scores, -top_k)[-top_k:]
+        if scores[i] >= score_threshold
+    ]
+    return sorted(classes, key=operator.itemgetter(1), reverse=True)
```
as above.

amj commented on raspicam/classify_capture.py:
```diff
@@ -14,14 +14,59 @@
+import tflite_runtime.interpreter as tflite
+import time
+
+Class = collections.namedtuple('Class', ['id', 'score'])
```
as above

amj commented on gstreamer/classify.py:
```diff
@@ -14,13 +14,17 @@
+import svgwrite
+import time
+
+Class = collections.namedtuple('Class', ['id', 'score'])
```
consider a different name :)

mtyka commented on pygame/classify_capture.py:
```diff
@@ -14,16 +14,61 @@
+                               {'device': device[0]} if device else {})
+      ])
+
+def input_size(interpreter):
```
Since you're not returning the full  tensor size but just the image size, i'd change the name of this helper to "input_image_size" or return the full tensor size

mtyka commented on pygame/classify_capture.py:
```diff
@@ -14,16 +14,61 @@
+    scale, zero_point = output_details['quantization']
+    return scale * (output_data - zero_point)
+
+def set_interpreter(interpreter, data):
```
This is an odd name since it invokes the interpreter. Why not "invoke_interpreter" ?
Honestly i'd just get rid of this helper entirely. The reshape really belongs with the buffer retriever step, not with the invoker step and interpreter.invoke() is already a clean call.

mtyka commented on gstreamer/classify.py:
```diff
@@ -35,6 +39,23 @@ def generate_svg(size, text_lines):
       dwg.add(dwg.text(line, insert=(10, y*20), fill='white', font_size='20'))
     return dwg.tostring()
 
+def output_tensor(interpreter):
```
Move outout_tensor() to common.py and merge with the other output_tensor(interpreter, i)
(and invoke it here with i=0).

mtyka commented on gstreamer/classify.py:
```diff
@@ -35,6 +39,23 @@ def generate_svg(size, text_lines):
+    """Returns dequantized output tensor."""
+    output_details = interpreter.get_output_details()[0]
+    output_data = np.squeeze(interpreter.tensor(output_details['index'])())
+    scale, zero_point = output_details['quantization']
```
add before:
```if quantization not in output_details: return output_data```
and elsewhere where dequantization is used.

mtyka commented on gstreamer/detect.py:
```diff
@@ -64,12 +67,41 @@ def generate_svg(src_size, inference_size, inference_box, objs, labels, text_lin
                         fill='none', stroke='red', stroke_width='2'))
     return dwg.tostring()
 
+def output_tensor(interpreter, i):
```
Merge with the other "output_tensor" and move to common.py

mtyka commented on pygame/classify_capture.py:
```diff
@@ -14,16 +14,61 @@
+    _, height, width, channels = interpreter.get_input_details()[0]['shape']
+    return width, height, channels
+
+def input_tensor(interpreter):
```
since ../pygame/  has detect and classify examples, consider using a ```common.py``` here

mtyka commented on raspicam/classify_capture.py:
```diff
@@ -14,14 +14,59 @@
+                               {'device': device[0]} if device else {})
+      ])
+
+def input_size(interpreter):
```
Even though there is only one example here i'd still factor these helpers into a common.py.

Ideally the common.py would be identical between the different sub-example-subfolders but i dont see an easy way
to share just one file without making the import more complicated.

mtyka commented:
Overall looks great, just a few minor comments! Thanks !

manderfeld commented:
PTAL @amj @mtyka 

General summary of changes:
- Updated README files
- Updated/added common.py files for all examples
- Moved common functions to common.py -- almost standard accross all examples
- Modified print statements to use .format to increase readability
- Unified string quotations (I used single quotes for strings, but this can be changed)

mtyka commented:
LGTM, Thanks!
